catalyst_voices/packages/libs/catalyst_cose/test/cose_sign_test.dart

```diff
@@ -2,47 +2,51 @@ import 'dart:convert';
 import 'dart:typed_data';
 
 import 'package:catalyst_cose/catalyst_cose.dart';
-import 'package:convert/convert.dart';
 import 'package:cryptography/cryptography.dart';
 import 'package:test/test.dart';
 
 void main() {
   group(CoseSign, () {
     const uuidV7 = '0193b535-7196-7cd1-84e6-ad9c316cf2d2';
-    late SimplePublicKey publicKey;
     late _SignerVerifier signerVerifier;
 
     setUp(() async {
       // Initialize the Ed25519 algorithm and generate a key pair
       final algorithm = Ed25519();
       final keyPair = await algorithm.newKeyPairFromSeed(List.filled(32, 0));
-      publicKey = await keyPair.extractPublicKey();
       signerVerifier = _SignerVerifier(algorithm, keyPair);
     });
 
     test('sign generates a valid COSE_SIGN structure', () async {
       final coseSign = await CoseSign.sign(
-        protectedHeaders: CoseHeaders.protected(
-          alg: const IntValue(CoseValues.eddsaAlg),
-          kid: hex.encode(publicKey.bytes),
-          contentType: const IntValue(CoseValues.jsonContentType),
-          contentEncoding: const StringValue(CoseValues.brotliContentEncoding),
-          type: const Uuid(uuidV7),
-          id: const Uuid(uuidV7),
-          ver: const Uuid(uuidV7),
-          ref: const SingleReferenceUuid(Uuid(uuidV7)),
-          template: const SingleReferenceUuid(Uuid(uuidV7)),
-          reply: const SingleReferenceUuid(Uuid(uuidV7)),
+        protectedHeaders: const CoseHeaders.protected(
+          contentType: IntValue(CoseValues.jsonContentType),
+          contentEncoding: StringValue(CoseValues.brotliContentEncoding),
+          type: Uuid(uuidV7),
+          id: Uuid(uuidV7),
+          ver: Uuid(uuidV7),
+          ref: SingleReferenceUuid(Uuid(uuidV7)),
```

Comment:
Doc ref, template, and reply can be either a single reference UUIDv7 or an array of exactly two UUIDv7, the first item representing the ID, the second item represents the Doc Ver.

Reply:
Yep, it's implemented like this. Unfortunately union types in dart are not that readable but it's either a single uuid v7 or an array of two.

Reply:
Could the type then be renamed to reflect that it isn't a single reference uuid? It seems to me that the name should reflect that the value can either a single value, or an array of two. Any name that improves readability will improve maintainability, it will also help reduce onboarding time for people wanting to learn how to use the package.

Reply:
You can think of a sealed class like an enum but with extra fields. A sealed class is a supertype and for each enum value there is a subclass. Here, the supertype was `ReferenceUuid` and this is what `ref` field is accepting.

There are two subclasses (enum values) for it: `SingleReferenceUuid` and `DoubleReferenceUuid`. If you wanted to assign a single uuid then you'd would use `SingleReferenceUuid`, if you needed to assign an array of two you would go for `DoubleReferenceUuid`.

In this specific example my intention was to assign a single uuid to the `ref` so I picked a subclass that corresponds to it. So yes, it's already possible to use either a single uuid or an array of two.

--

As I learned more about the context how it would be used I have come up with another structure that represents it better.

```dart

final class ReferenceUuid extends Equatable {
  /// The referenced entity uuid.
  final Uuid id;

  /// The version of the referenced entity.
  final Uuid? ver;

  /// The default constructor for the [ReferenceUuid].
  const ReferenceUuid({
    required this.id,
    this.ver,
  });
```

My understanding is that we are referring to an id of some entity (document, etc) and optionally we might specify the version of it. Therefore a structure that has a required `id` parameter and optional `ver` parameter reflects it better because it properly names the fields that are needed. The structure is smart enough to encode it as cbor list if `ver` is given or just as cbor bytes if `ver` is not given.